app/services/pdf_layout_engine.rb

```diff
@@ -0,0 +1,44 @@
+module Services
+    class PDFLayoutEngine
+        def layout(object, component, document, width, height, left, top)
+            if !document then
+                document = Prawn::Document.new(:page_size => [height, width],
+                                      :page_layout => :landscape,
+                                      :margin => 0)
+                documentfont_size = 18
+            end
+            
+            
+            # Calculate actual sizes from proportional sizes
+            if component.layout then
+                w = width * component.layout.width
+                h = height * component.layout.height
+                l = left + component.layout.left * width
+                t = (top - height)  +  component.layout.top * height
+            else
+                # if not layout use containing layout
+                w = width
+                h = height
+                l = left
+                t = top
+            end
+            if component.respond_to?(:content) && component.content &&
+                    !component.content.empty? then
+                component.content.each do |inner_component|
+                    layout(object, inner_component, document, w, h, l, t )
+                end
+            else
+                if component.template_name == 'wizard/components/qr' then
```

Comment:
it would be cool, if the component could take care of rendering itself rather than having some if elses here.